WIP

bebehei opened the pull request:
Issues to fix:

- [x] include graphite module
  - Work [has been done](https://github.com/bebehei/icinga2-docker/commit/82e74b5c496b5d93984f8a47f38700e357aa73ca), but I dislike the official graphite module, the other one by [findmypast](https://github.com/findmypast/icingaweb2-module-graphite) may be better.
- [x] some refactoring
- [x] docs improvement
- [x] navigation items are lost after creation of new container
  - Resolved: I forgot to mount volume on `/etc/icingaweb2`
- [x] module information on status pages is not
  - fix: add module path declaration and add `/etc/icingaweb2/modules`
- [x] unhardcode passwords (#40)

bebehei commented:
Ok, to clearify this. I'll push many changes here. If you could review them, that'll be great.

But please don't merge them by now. I will add more (and bigger) commits in the next days/weeks. If you would merge these all one by one, that would give a very confusing commit history.

jjethwa commented on content/opt/run:
```diff
@@ -219,5 +212,5 @@ echo "==================================================================="
 fi
 
-echo "Starting Supervisor.  You can safely CTRL-C and the container will continue to run with or without the -d (daemon) option"
+echo "Starting Supervisor."
```
I added this note  to let users know it was OK to break out as there seemed to be a lot of confusion for new users.  I think we should keep it in for now 😛

bebehei replied:
I experienced differently. Every time i hit Ctrl+C it stops.

Probably this is, because docker run bridges the sequence to the supervisor.

using docker 1.12

bebehei replied:
I've heard docker minor versions are often not compatible to each other. In fact I'm quite new to docker software and I have only used the 1.12 version yet. But I have to say, that this is sadly not true in 1.12.

```
[bebe:~/code/icinga2-docker] devtest ± docker run --rm -ti jordan/icinga2
[...]
Starting Supervisor.  You can safely CTRL-C and the container will continue to run with or without the -d (daemon) option
^C
[bebe:~/code/icinga2-docker] devtest ± docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[bebe:~/code/icinga2-docker] devtest ±
```

jjethwa replied:
I think it's because you're running the container with the -i interactive flag so you're actually sending the break to supervisor and not the host.

jjethwa replied:
Just tested it out and it does appear to be the case.  To detach from an interactive session, you would need to use sequence of CTRL + P and CTRL + Q sequence.  

Unless there's a reason to start a container in interactive mode, I think the general recommendation is to run with the -d daemon flag or since this container uses supervisor to run the services, you can run it without the -d daemon flag and just escape out

jjethwa commented:
Hi @bebehei 

Thanks!  I really appreciate all the work you've done.  I've added you as a collaborator.   Docker Hub is set up to build master -> latest, dev -> latest, and tagged as their tag numbers.  For this PR, do you mind submitting it to the dev branch?  Then once it's checked out from the Docker hub dev container, you can submit a PR from dev to master?

Unfortunately, I don't have tests set up externally.  My tests run through basic scenarios using the environment variables and volumes (empty + pre-existing configs, database, etc)

bebehei commented:
> Thanks! I really appreciate all the work you've done. I've added you as a collaborator. Docker Hub is set up to build master -> latest, dev -> latest, and tagged as their tag numbers. For this PR, do you mind submitting it to the dev branch? Then once it's checked out from the Docker hub dev container, you can submit a PR from dev to master?

Thanks, I appreciate to be a "Member" of this repository. But actually, I prefer working on in this PR. This makes force pushing more convenient than doing it on an official branch. Or are there any improvements?

> Unfortunately, I don't have tests set up externally. My tests run through basic scenarios using the environment variables and volumes (empty + pre-existing configs, database, etc)

Do you have tests? If yes, where are these located? Currently, I just test it manuall with every change.

jjethwa commented:
Hi @bebehei 

Agreed on using PRs for work like you have been.  Added you as a collaborator so you could merge PRs and perform other repo tasks as needed as I don't want to hold you back 😄

Right now the tests are not in a good shape to share as they are hardcoded to use my local infra and are not using a proper framework like Selenium.  We should look into setting up CircleCI, Travis, etc to build the container and run tests against them in the future.

bebehei commented:
Hi @jjethwa,

I would like to merge this now. Could you please review this? This might have a bug I have not hit (yet).

- refactored code:
  - split everything related to setting up the service into a separate file
  - many simplifications
- initfile is not neccessary anymore. All changes, which are changeable with an ENV-Variable are now safe to reexecute all the time.
- added support for login-credentials (067fb3d76f0b18733de578669279e8e08d1797da)
- added graphite web support (7916a300d5c1d6a891d20876868a8e4f31ca3869)

things not done yet: (PRs will follow)
- more detailed README
- http -> https redirection
- detailed `docker-compose.yml` with graphite and maybe support for external mysql DB

bebehei commented:
And as you may know: `docker pull bebehei/icinga2:smaller-install` gives you this PR-branch prebuilt by docker hub.

bebehei commented:
Except those small showstoppers, I'd like to finalise it and merge this heavy PR.

I'd like to get Feedback, of course. ;-)

bebehei commented on README.md:
```diff
@@ -1,79 +1,141 @@
+| /var/lib/mysql | rw | MySQL Database |
+| /var/lib/icinga2 | rw | Icinga2 Data |
+| /var/log/apache2 | rw | logfolder for apache2 (not neccessary) |
+| /var/log/icinga2 | rw | logfolder for icinga2 (not neccessary) |
```
TODO: Have all the log-dirs been chowned by the setup-scripts?

bebehei replied:
fixed this now, too.

jjethwa commented:
Hi @bebehei 

Sorry, I took a long weekend away from my computer 😄 

Give me a few days to review and test since this is a huge PR hehe

jjethwa commented on content/opt/setup/icinga2:
```diff
@@ -0,0 +1,59 @@
+  GRANT SELECT, INSERT, UPDATE, DELETE, DROP, CREATE VIEW, INDEX, EXECUTE ON icinga.* TO 'icinga'@'localhost' IDENTIFIED BY '${ICINGA_PASSWORD}';
+END
+
+mysql icinga < /usr/share/icinga2-ido-mysql/schema/mysql.sql >> /var/log/icinga2/mysql-schema.log 2>&1
```
Is there a reason to run the schema again?  Is this to handle icinga2-ido-mysql upgrades?

bebehei replied:
The key thought on this change had been:

On a docker-container you're able to configure your stuff via `ENV`-variables. And you save the stuff via volumes. So during boot, the docker-container should process the variables and change the volume data to represent the ENV-settings. The current master uses a logfile approach and changes everything on first container boot and touches a lockfile.

While volumes are persistent over container creation, `ENV`s are not. This leads to a new problem: If you change your ENVs, the volume data does not change automatically. So, my intention is to run the complete setup on every boot, but only **change neccessary values** without overwriting other stuff (which may be created manually by the user). This is why, I removed the sed-calls with `ini_set`, etc.

I created this initial note, to remember, that I haven't tested this actively. It worked, but I didn't know if it worked correctly.

As you mentioned, schema upgrades are also another topic, which this image still requires manual interaction.

bebehei replied:
Tested it: MySQL just stops after the first failing command, because there are duplicate columns in DB-Table. And first failing command is the first command, if the DB is already created and filled. So nothing bad will happen. I would "mark" this as done.

jjethwa commented on content/etc/icingaweb2/config.ini:
```diff
@@ -0,0 +1,10 @@
+show_stacktraces = "1"
+config_backend = "db"
+config_resource = "icingaweb_db"
+module_path = "/usr/share/icingaweb2/modules:/etc/icingaweb2/modules"
```
Agreed.  The RC version of Icinga Web 2 had the modules installed under /etc/icingaweb2/modules instead of /use/share/icingaweb2/modules.  I believe the RPM version still does that, but the DEB version was updated later to install the modules under /use/share/icingaweb2/modules.  Since this container is already in the wild and being used by users in Production, we'd need to make sure that any changes don't break the current set up. 

Perhaps install the module to their correct path under /usr/share/icingaweb2/modules then symlink from /etc/icingaweb2/modules to /usr/share/icingaweb2/modules ?

bebehei replied:
There is the `module_path` for icingaweb2. It just works like a normal path-variable and tells icignacli/icingaweb2 where to search for plugins. I think we have to define 3 locations:
- for self-installed modules from the user -> `/etc/icingaweb2/modules-user` (on mounted volume)
- for the ones installed from docker-container -> `/opt/icingaweb2/modules` (not on volume) 
- for the ones by debian-packages -> `/usr/share/icingaweb2/modules`

jjethwa replied:
Sounds good!  Right now, the only example of a docker from the docker-container would be https://github.com/findmypast/icingaweb2-module-graphite right?

bebehei replied:
No, the dockerfile currently uses the official module. I thought about using findmypast's module, but it has got some technical disadvantages.

jjethwa replied:
I use the findmypast's module, but you also need to use the official module to send the data to carbon

jjethwa commented on content/opt/setup/graphite:
```diff
@@ -0,0 +1,25 @@
+
+else
+	# Actively disable graphite, to not hit any weird bugs
+	icinga2 feature disable graphite
```
No need to disable graphite as it is not enabled by default, right?

The container output might be a bit confusing for users:
```
critical/cli: Cannot disable feature 'graphite'. Target file '/etc/icinga2/features-enabled/graphite.conf' does not exist.
critical/cli: Cannot disable feature(s): graphite
=> Copying fresh config-files for /etc/icingaweb2
/opt/setup/icingaweb2: line 31: ini: command not found
ERROR: Icinga\Exception\ConfigurationError in /usr/share/php/Icinga/Application/Modules/Manager.php:318 with message: Cannot disable module "graphite". Module is not installed.
```

bebehei replied:
Yes, this is a bit confusing. But just piping the output to `/dev/null` may pipe more than this. IMHO this should be icinga2's job to handle it.

bebehei replied:
Forgot to mention: I created a bug report, but the dev's haven't answered it yet. I hope they'll acknowledge it as a bug https://dev.icinga.com/issues/13903 . If I'll change something depends on their answer.

jjethwa replied:
Piping to /dev/null is a good idea.  I have a feeling the behaviour from the icinga2 CLI command is expected

bebehei replied:
> No need to disable graphite as it is not enabled by default, right?

As mentioned in another comment, I want the image to respect the ENVs. So if you would disable graphite with the ENV, graphite should be disabled without manual interaction, too.

> Piping to /dev/null is a good idea. I have a feeling the behaviour from the icinga2 CLI command is expected

I expect the dev's answer in the next few days also I expect to not be finished with QA before that.

bebehei replied:
Added a commit to only disable it, if it's actually enabled.

bebehei commented on content/opt/setup/icingaweb2-director:
```diff
@@ -0,0 +1,18 @@
+else
+	cat /etc/icingaweb2/modules/director/director.api >> /etc/icinga2/conf.d/api-users.conf
+fi
+sed -i 's,directorapi,'${DIRECTOR_API_PASSWORD}',g' /etc/icinga2/conf.d/api-users.conf
```
this does not work after 2nd boot.... 👎

bebehei replied:
this should be now fixed in latest revision

jjethwa commented:
Wow you're on a rampage 😄

bebehei commented:
> Wow you're on a rampage 😄

![yes](http://i.giphy.com/10pR6dUXHBpZSg.gif)

I wanna set this up in prod. I have to be on rampage.

jjethwa commented:
Great!  I'll pull the latest container and continue with the testing.  Thanks 😄

bebehei commented:
The last fixup commits from 9 hours ago, were unspectacular commits. Currently preparing icinga2 director. I may need it tomorrow at the office

jjethwa commented:
I should be done with my tests by tomorrow hopefully.

Once things are stable, we need to work out a plan for pushing this master to a new versioned branch on official Icinga2 docker repo and making deb container builds available via their official Docker Hub account.   The current CentOS container will continue to be maintained, but the deb one will get all latest updates as it's coming from this repo.

bebehei commented:
Official images? AFAIK there are only "official" docker-images for development!?

jjethwa commented:
That's correct.  The "official" CentOS image will continue to be for development, but the deb image will be based off this repo since it's already being used in production by a lot of users 😄

bebehei commented:
Oh, wait. So you're telling me, that you're the maintainer of the **official** icinga2 docker images... I did not see that coming. 😂

jjethwa commented:
lol well not exactly!  I talked to some of the team last year about helping out since the container is not 100% maintained. They work mostly with Vagrant or regular installs.    I was too busy for the past few months, but now that you've been helping out so much, I think we can definitely help them out ☺️

bebehei commented:
So, I added some fixes again. Additionally, I want to fix correct support on module_path. I will work on this in a few hours again. The code should then contain no known issues anymore.

bebehei commented:
> I want to fix correct support on module_path. I will work on this in a few hours again.

Fixed. I'll test it the next days, but I guess this is now safe to merge.

bebehei commented:
I looked on the issues in the issue tracker and I have to conclude: By merging this PR, you fix everything in this repo, except my feature request ;-)

jjethwa commented:
Thanks @bebehei 

Re-running the tests.  Let's put a freeze on any new changes to the branch/PR unless it's a major bug so we can get it fully tested and merged.  😛

bebehei commented:
Yes! For the last days, I'm doing this, too.

jjethwa commented:
@bebehei Looks good on my end and all my test cases seem to work.  I did have to update one old config file under /etc/icinga2 but that was my fault for not keeping it updated.  Thanks again for all of your hard work! 😄

bebehei commented:
Thanks. :stunned:
